### PR TITLE
Rework ShowPage and IndexPage in Shop/Product to DataAttributes

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 1]
```diff
@@ -24,31 +24,31 @@ public function getRouteName(): string
...
         return count($products);
     }
 
     public function getFirstProductNameFromList(): string
     {
-        $productsList = $this->getDocument()->find('css', '#products');
+        $productsList = $this->getDocument()->find('css', '[data-test-products]');
 
-        return $productsList->find('css', '.card:first-child .content > a')->getText();
+        return $productsList->find('css', '[data-test-product]:first-child [data-test-product-content] > a')->getText();
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: IMO you should add attribtute to this link  instead use '> a'

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 2]
```diff
@@ -24,31 +24,31 @@ public function getRouteName(): string
...
 
-        return $productsList->find('css', '.card:last-child .content > a')->getText();
+        return $productsList->find('css', '[data-test-product]:last-child [data-test-product-content] > a')->getText();
     }
 
     public function search(string $name): void
     {
         $this->getDocument()->fillField('criteria_search_value', $name);
-        $this->getDocument()->pressButton('Search');
+        $this->getDocument()->find('css', '[data-test-search]')->getParent()->submit();
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: Can you add this attribute on button and have :
```suggestion
        $this->getDocument()->find('css', '[data-test-search]')->submit();
```

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 3]
```diff
@@ -58,35 +58,37 @@ public function sort(string $orderNumber): void
...
-        return null !== $this->getDocument()->find('css', sprintf('.sylius-product-name:contains("%s")', $productName));
+        $element = $this->getDocument()->find('css', sprintf('[data-test-product-name="%s"]', $productName));
+
+        return ($element !== null) ? true : false;
     }
 
     public function isEmpty(): bool
     {
-        return false !== strpos($this->getDocument()->find('css', '.message')->getText(), 'There are no results to display');
+        return false !== strpos($this->getDocument()->find('css', '[data-test-flash-message] p')->getText(), 'There are no results to display');
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: Can you add attritbute on this paragraph instad use 'p' ?

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_options.html.twig [hunk 4]
```diff
@@ -1,3 +1,3 @@
 {% for option_form in form.cartItem.variant %}
-    {{ form_row(option_form, { 'attr': { 'data-option': option_form.vars.name } }) }}
+    {{ form_row(option_form, sylius_test_form_attribute('item-variant-'~option_form.vars.name)|merge({ 'data-option': option_form.vars.name }) ) }}
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_options.html.twig]:
```suggestion
    {{ form_row(option_form, sylius_test_form_attribute('item-variant-' ~ option_form.vars.name)|merge({ 'data-option': option_form.vars.name }) ) }}
```

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php [hunk 5]
```diff
@@ -61,10 +61,10 @@ public function addToCartWithVariant(string $variant): void
 
     public function addToCartWithOption(ProductOptionInterface $option, string $optionValue): void
     {
-        $select = $this->getDocument()->find('css', sprintf('select#sylius_add_to_cart_cartItem_variant_%s', $option->getCode()));
+        $select = $this->getDocument()->find('css', sprintf('[data-test-item-variant-%s]', $option->getCode()));
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: All data like code, name, etc we should put in attribute value for example: ``[data-test-item-variant="%s"]``
here and in other places.

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 6]
```diff
@@ -24,31 +24,31 @@ public function getRouteName(): string
 
     public function countProductsItems(): int
     {
-        $productsList = $this->getDocument()->find('css', '#products');
+        $productsList = $this->getDocument()->find('css', '[data-test-products]');
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: Use `getElement` here

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 7]
```diff
@@ -58,35 +58,37 @@ public function sort(string $orderNumber): void
 
     public function clearFilter(): void
     {
-        $this->getDocument()->clickLink('Clear');
+        $this->getDocument()->find('css', '[data-test-clear]')->click();
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: Use `getElement`

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php (hunk 1 above)
Comment (Tomanhez) [reply src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: If i good remember we talking that good practice will be use data-attributes and if we need html attributes, we go to working without css class.

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 8]
```diff
@@ -24,31 +24,31 @@ public function getRouteName(): string
...
 
-        return $productsList->find('css', '.card:last-child .content > a')->getText();
+        return $productsList->find('css', '[data-test-product]:last-child [data-test-product-content] > a')->getText();
     }
 
     public function search(string $name): void
     {
         $this->getDocument()->fillField('criteria_search_value', $name);
-        $this->getDocument()->pressButton('Search');
+        $this->getDocument()->find('css', '[data-test-search]')->submit();
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: You can use here `getElement` method

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 9]
```diff
@@ -58,35 +58,37 @@ public function sort(string $orderNumber): void
...
     public function clearFilter(): void
     {
-        $this->getDocument()->clickLink('Clear');
+        $this->getElement('clear')->click();
     }
 
     public function isProductOnList(string $productName): bool
     {
-        return null !== $this->getDocument()->find('css', sprintf('.sylius-product-name:contains("%s")', $productName));
+        $element = $this->getDocument()->find('css', sprintf('[data-test-product-name="%s"]', $productName));
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: You can use here `getElement` method

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 10]
```diff
@@ -58,35 +58,37 @@ public function sort(string $orderNumber): void
...
+
+        return ($element !== null) ? true : false;
     }
 
     public function isEmpty(): bool
     {
-        return false !== strpos($this->getDocument()->find('css', '.message')->getText(), 'There are no results to display');
+        return false !== strpos($this->getDocument()->find('css', '[data-test-flash-message]')->getText(), 'There are no results to display');
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: You can use here `getElement` method

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 11]
```diff
@@ -58,35 +58,37 @@ public function sort(string $orderNumber): void
...
     }
 
     public function getProductPrice(string $productName): string
     {
-        $container = $this->getDocument()->find('css', sprintf('.sylius-product-name:contains("%s")', $productName))->getParent();
+        $element = $this->getDocument()->find('css', sprintf('[data-test-product-name="%s"]', $productName));
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: You can use here `getElement` method

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 12]
```diff
@@ -58,35 +58,37 @@ public function sort(string $orderNumber): void
...
 
-        return $container->find('css', '.sylius-product-price')->getText();
+        return $element->getParent()->find('css', '[data-test-product-price]')->getText();
     }
 
     public function isProductOnPageWithName(string $name): bool
     {
-        return null !== $this->getDocument()->find('css', sprintf('.content > a:contains("%s")', $name));
+        return null !== $this->getDocument()->find('css', sprintf('[data-test-product-name="%s"]', $name));
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: You can use here `getElement` method

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php [hunk 13]
```diff
@@ -61,10 +61,10 @@ public function addToCartWithVariant(string $variant): void
 
     public function addToCartWithOption(ProductOptionInterface $option, string $optionValue): void
     {
-        $select = $this->getDocument()->find('css', sprintf('select#sylius_add_to_cart_cartItem_variant_%s', $option->getCode()));
+        $select = $this->getDocument()->find('css', sprintf('[data-test-item-variant="%s"]', $option->getCode()));
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: You can use here `getElement` method

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php [hunk 14]
```diff
@@ -121,18 +120,21 @@ public function getCurrentVariantName(): string
 
     public function getName(): string
     {
-        return $this->getElement('name')->getText();
+        return $this->getDocument()->find('css', '[data-test-product-name]')->getText();
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: You can use here `getElement` method

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php [hunk 15]
```diff
@@ -146,7 +148,7 @@ public function hasProductInAssociation(string $productName, string $productAsso
 
         Assert::notNull($products);
 
-        return null !== $products->find('css', sprintf('.sylius-product-name:contains("%s")', $productName));
+        return $productName === $products->find('css', sprintf('[data-test-product-name="%s"]', $productName))->getText();
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: You can use here `getElement` method

Comment (Tomanhez) [reply src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: Here I can't because if I use getElement and return products, next if I try use getElement on products I get error.

Screenshot 2019-12-20 at 07 27 18

File: behat.yml.dist [hunk 16]
```diff
@@ -42,8 +42,6 @@ default:
                                     - "no-sandbox"
                             extra_capabilities:
                                 unexpectedAlertBehaviour: accept
-                                goog:chromeOptions:
-                                    w3c: false # https://github.com/Sylius/Sylius/issues/10561
```
Comment (AdamKasp) [inline behat.yml.dist]: you shouldn't delete it .

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 17]
```diff
@@ -58,35 +59,41 @@ public function sort(string $orderNumber): void
...
     public function clearFilter(): void
     {
-        $this->getDocument()->clickLink('Clear');
+        $this->getElement('clear')->click();
     }
 
     public function isProductOnList(string $productName): bool
     {
-        return null !== $this->getDocument()->find('css', sprintf('.sylius-product-name:contains("%s")', $productName));
+        try{
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: why did you change it to try/catch block?

Comment (Tomanhez) [reply src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: when I replace getDocument() to getElement() and element will not exist, getElement throw exception, we can use hasElement if all element is defined like hasElement('product') but if we try hasElement('product', ['%productName% => $productName']) we get error

**Comment by GSadee** [comment]
Thank you, Tomasz! :1st_place_medal:
